### ISSUE TITLE
prefer-lodash-typecheck does not catch typeof configuration[name] === 'undefined' 

### ISSUE THREAD

**Issue by gajus** [wix/eslint-plugin-lodash#10]
should suggest to use https://lodash.com/docs#isUndefined

The specific subject where it did not suggest to replace it with `_.isUndefined`:

```js
import _ from 'lodash';

/**
 * @typedef CSSModules~Options
 * @see {@link https://github.com/gajus/react-css-modules#options}
 * @property {boolean} allowMultiple
 * @property {boolean} errorWhenNotFound
 */

/**
 * @param {CSSModules~Options} userConfiguration
 * @return {CSSModules~Options}
 */
export default (userConfiguration = {}) => {
    let configuration;

    configuration = {
        allowMultiple: false,
        errorWhenNotFound: true
    };

    _.forEach(userConfiguration, (value, name) => {
        if (typeof configuration[name] === 'undefined') {
            throw new Error('Unknown configuration property "' + name + '".');
        }

        if (typeof value !== 'boolean') {
            throw new Error('"' + name + '" property value must be a boolean.');
        }

        configuration[name] = value;
    });

    return configuration;
};
```

**Comment by mik01aj** [comment]
See https://github.com/wix/eslint-plugin-lodash3/issues/5.

**Comment by gajus** [comment]
@mik01aj The example that you have provided has extremely limited use case.

It is only relevant when checking for variables in the inherited scope (globals, such as in the context of browser scripts).

Fair to say that the default behaviour should be to enforce the use of `_.isUndefined` through ESLint rules and use `/* eslint-disable lodash3/prefer-lodash-typecheck */` in such very specific cases.

You will never get such an error when the variable declaration is present, e.g. when checking function parameter presence.

```
(a) => {
    // Does not produce an error.
    _.isUndefined(a);
};
```

**Comment by mik01aj** [comment]
Agreed. :+1:

**Comment by gajus** [comment]
@ganimomer any input?

**Comment by ganimomer** [comment]
The intent of the rule is to prefer Lodash versions of type checking, as long as they have the exact same meaning and behavior in all cases.
In the case of _.isUndefined, it is not the exact same is `typeof(a) === 'undefined`.
I suppose the rule *could* use the context to check if either the expression is a variable that exists or isn't a variable (i.e., `typeof a.b === 'undefined'` or `a => typeof(a) === 'undefined') 
but the rule definitely shouldn't have 'false positives' like you're suggesting, where the replacement shouldn't happen and the rule should be turned off.
We could get around to implementing that soon, or anyone could submit a PR if you like.

**Comment by mik01aj** [comment]
I think that this could be configurable, so that the rule wouldn't warn about typechecks with undefined by default, but you could have a `alsoForUndefined` flag that would enable this.

**Comment by ganimomer** [comment]
I don't agree.
**Under no configuration should any rule ever report in a case where fixing it would cause an error.**
However, like I said, it **can** only report `typeof <expression> === 'undefined'` where fixing is possible.
ESLint enables rules to check which variables are declared with [`context.getDeclaredVariables()`](http://eslint.org/docs/developer-guide/working-with-rules#the-context-object)

**Comment by mik01aj** [comment]
@ganimomer so how about triggering a warning only when using it as `typeof <declared-variable> === 'undefined'`? In this case, lodash typecheck is equivalent.

**Comment by ganimomer** [comment]
That's the idea. :)
Well, that and checking any expression that isn't a literal.
technically `typeof a.b === 'undefined'` will behave the same way as `_.isUndefined(a.b)`.
v0.4.0 we released yesterday fixed *that* part, but still hasn't fixed the part where it's a declared literal, since that's a bit more complicated.

**Comment by mik01aj** [comment]
Great, thanks! :+1:

**Comment by ganimomer** [comment]
Technically, this issue is resolved. I'm closing it and opening a separate issue for `typeof <declared-variable>`.
